File: results/2026-04-02_163146/tasks/02-log-file-analyzer/default-opus/generated-code/test_log_analyzer.js
#!/usr/bin/env node
/**
 * Tests for log_analyzer.js — written following red/green TDD methodology.
 *
 * Each describe block corresponds to a TDD cycle:
 *   1. RED:   Write the test first (it fails because the code doesn't exist).
 *   2. GREEN: Write the minimum implementation to make it pass.
 *   3. REFACTOR: Clean up without breaking tests.
 *
 * Uses Node.js built-in test runner (node:test) and assertions (node:assert).
 * Run with: node --test test_log_analyzer.js
 */

const { describe, it } = require("node:test");
const assert = require("node:assert/strict");
const path = require("node:path");
const fs = require("node:fs");
const os = require("node:os");

const {
  parseSyslogLine,
  parseJsonLine,
  parseLine,
  filterErrorsAndWarnings,
  buildFrequencyTable,
  parseLogFile,
  formatTable,
  analyzeLogFile,
} = require("./log_analyzer");

const FIXTURES = path.join(__dirname, "fixtures");

// ---------------------------------------------------------------------------
// Cycle 1: Parse syslog-style log lines
// ---------------------------------------------------------------------------
describe("parseSyslogLine", () => {
  it("parses an ERROR line", () => {
    const line =
      "2024-01-15 08:23:45 ERROR [auth] Failed login attempt for user admin";
    const result = parseSyslogLine(line);
    assert.deepStrictEqual(result, {
      timestamp: "2024-01-15 08:23:45",
      level: "ERROR",
      source: "auth",
      message: "Failed login attempt for user admin",
    });
  });

  it("parses a WARNING line", () => {
    const line = "2024-01-15 09:10:02 WARNING [disk] Disk usage above 85%";
    const result = parseSyslogLine(line);
    assert.equal(result.level, "WARNING");
    assert.equal(result.source, "disk");
  });

  it("parses an INFO line", () => {
    const line = "2024-01-15 10:00:00 INFO [app] Server started on port 8080";
    const result = parseSyslogLine(line);
    assert.equal(result.level, "INFO");
  });

  it("returns null for non-matching lines", () => {
    assert.equal(parseSyslogLine("random garbage"), null);
  });

  it("returns null for empty string", () => {
    assert.equal(parseSyslogLine(""), null);
  });

  it("handles leading/trailing whitespace", () => {
    const line = "  2024-01-15 08:23:45 ERROR [auth] Trimmed  ";
    const result = parseSyslogLine(line);
    assert.equal(result.message, "Trimmed");
  });
});

// ---------------------------------------------------------------------------
// Cycle 2: Parse JSON-structured log lines
// ---------------------------------------------------------------------------
describe("parseJsonLine", () => {
  it("parses a valid JSON error entry", () => {
    const line = JSON.stringify({
      timestamp: "2024-01-15T08:35:22Z",
      level: "ERROR",
      source: "database",
      message: "Connection timeout after 30s",
    });
    const result = parseJsonLine(line);
    assert.equal(result.level, "ERROR");
    assert.equal(result.source, "database");
    // ISO timestamp normalised to YYYY-MM-DD HH:MM:SS
    assert.equal(result.timestamp, "2024-01-15 08:35:22");
  });

  it("upper-cases the level field", () => {
    const line = JSON.stringify({
      timestamp: "2024-01-15T09:10:05Z",
      level: "warning",
      source: "memory",
      message: "Heap usage above 90%",
    });
    assert.equal(parseJsonLine(line).level, "WARNING");
  });

  it("returns null for invalid JSON", () => {
    assert.equal(parseJsonLine('{"incomplete'), null);
  });

  it("returns null when required fields are missing", () => {
    const line = JSON.stringify({
      timestamp: "2024-01-15T09:00:00Z",
      level: "ERROR",
      message: "missing source",
    });
    assert.equal(parseJsonLine(line), null);
  });

  it("returns null for non-JSON lines", () => {
    assert.equal(
      parseJsonLine("2024-01-15 08:23:45 ERROR [auth] Not JSON"),
      null
    );
  });

  it("returns null for empty string", () => {
    assert.equal(parseJsonLine(""), null);
  });
});

// ---------------------------------------------------------------------------
// Cycle 3: Unified line parser (tries JSON then syslog)
// ---------------------------------------------------------------------------
describe("parseLine", () => {
  it("parses syslog lines", () => {
    const line = "2024-01-15 08:23:45 ERROR [auth] Login failed";
    assert.equal(parseLine(line).source, "auth");
  });

  it("parses JSON lines", () => {
    const line = JSON.stringify({
      timestamp: "2024-01-15T10:00:00Z",
      level: "ERROR",
      source: "api",
      message: "Rate limit",
    });
    assert.equal(parseLine(line).source, "api");
  });

  it("returns null for garbage", () => {
    assert.equal(parseLine("not a log line"), null);
  });
});

// ---------------------------------------------------------------------------
// Cycle 4: Filter errors and warnings
// ---------------------------------------------------------------------------
describe("filterErrorsAndWarnings", () => {
  it("keeps only ERROR and WARNING entries", () => {
    const entries = [
      { level: "ERROR", timestamp: "t1", source: "a", message: "m1" },
      { level: "INFO", timestamp: "t2", source: "b", message: "m2" },
      { level: "WARNING", timestamp: "t3", source: "c", message: "m3" },
      { level: "DEBUG", timestamp: "t4", source: "d", message: "m4" },
    ];
    const result = filterErrorsAndWarnings(entries);
    assert.equal(result.length, 2);
    assert.equal(result[0].level, "ERROR");
    assert.equal(result[1].level, "WARNING");
  });

  it("returns empty array when no matches", () => {
    const entries = [
      { level: "INFO", timestamp: "t1", source: "a", message: "m1" },
    ];
    assert.deepStrictEqual(filterErrorsAndWarnings(entries), []);
  });

  it("returns empty array for empty input", () => {
    assert.deepStrictEqual(filterErrorsAndWarnings([]), []);
  });
});

// ---------------------------------------------------------------------------
// Cycle 5: Build frequency table with first/last timestamps
// ---------------------------------------------------------------------------
describe("buildFrequencyTable", () => {
  it("counts duplicate entries", () => {
    const entries = [
      {
        level: "ERROR",
        timestamp: "2024-01-15 08:23:45",
        source: "auth",
        message: "Login failed",
      },
      {
        level: "ERROR",
        timestamp: "2024-01-15 08:40:15",
        source: "auth",
        message: "Login failed",
      },
      {
        level: "ERROR",
        timestamp: "2024-01-15 10:15:00",
        source: "auth",
        message: "Login failed",
      },
    ];
    const freq = buildFrequencyTable(entries);
    assert.equal(freq.length, 1);
    assert.equal(freq[0].count, 3);
  });

  it("tracks first and last occurrence timestamps", () => {
    const entries = [
      {
        level: "ERROR",
        timestamp: "2024-01-15 08:23:45",
        source: "auth",
        message: "Login failed",
      },
      {
        level: "ERROR",
        timestamp: "2024-01-15 10:15:00",
        source: "auth",
        message: "Login failed",
      },
    ];
    const freq = buildFrequencyTable(entries);
    assert.equal(freq[0].first_seen, "2024-01-15 08:23:45");
    assert.equal(freq[0].last_seen, "2024-01-15 10:15:00");
  });

  it("groups by source and message", () => {
    const entries = [
      {
        level: "ERROR",
        timestamp: "t1",
        source: "auth",
        message: "Login failed",
      },
      {
        level: "WARNING",
        timestamp: "t2",
        source: "disk",
        message: "Low space",
      },
      {
        level: "ERROR",
        timestamp: "t3",
        source: "auth",
        message: "Login failed",
      },
    ];
    const freq = buildFrequencyTable(entries);
    assert.equal(freq.length, 2);
    // Highest count first
    assert.equal(freq[0].count, 2);
    assert.equal(freq[0].source, "auth");
  });

  it("sorts by count descending", () => {
    const entries = [
      { level: "ERROR", timestamp: "t1", source: "a", message: "m1" },
      { level: "ERROR", timestamp: "t2", source: "b", message: "m2" },
      { level: "ERROR", timestamp: "t3", source: "b", message: "m2" },
      { level: "ERROR", timestamp: "t4", source: "b", message: "m2" },
    ];
    const freq = buildFrequencyTable(entries);
    assert.equal(freq[0].count, 3);
    assert.equal(freq[1].count, 1);
  });

  it("returns empty array for empty input", () => {
    assert.deepStrictEqual(buildFrequencyTable([]), []);
  });
});

// ---------------------------------------------------------------------------
// Cycle 6: Parse a full log file from disk
// ---------------------------------------------------------------------------
describe("parseLogFile", () => {
  it("parses sample.log (mixed syslog + JSON)", () => {
    const entries = parseLogFile(path.join(FIXTURES, "sample.log"));
    assert.ok(entries.length > 0);
    const sources = new Set(entries.map((e) => e.source));
    assert.ok(sources.has("auth")); // syslog entry
    assert.ok(sources.has("database")); // JSON entry
  });

  it("parses json_only.log", () => {
    const entries = parseLogFile(path.join(FIXTURES, "json_only.log"));
    assert.equal(entries.length, 4);
    assert.ok(entries.every((e) => e.source === "cache"));
  });

  it("parses syslog_only.log", () => {
    const entries = parseLogFile(path.join(FIXTURES, "syslog_only.log"));
    assert.equal(entries.length, 5);
  });

  it("skips unparseable lines gracefully", () => {
    const entries = parseLogFile(path.join(FIXTURES, "malformed.log"));
    // 6 total lines: 2 garbage + 1 broken JSON = 3 skipped, 3 valid
    assert.equal(entries.length, 3);
  });

  it("returns empty array for empty file", () => {
    const entries = parseLogFile(path.join(FIXTURES, "empty.log"));
    assert.deepStrictEqual(entries, []);
  });

  it("throws for missing file with a clear message", () => {
    assert.throws(
      () => parseLogFile("/nonexistent/path/missing.log"),
      (err) => {
        assert.ok(err instanceof Error);
        assert.ok(err.message.includes("missing.log"));
        return true;
      }
    );
  });
});

// ---------------------------------------------------------------------------
// Cycle 7: Format human-readable table
// ---------------------------------------------------------------------------
describe("formatTable", () => {
  it("shows a message for empty input", () => {
    assert.equal(formatTable([]), "No error or warning entries found.");
  });

  it("includes column headers", () => {
    const freq = [
      {
        level: "ERROR",
        source: "auth",
        message: "fail",
        count: 1,
        first_seen: "2024-01-01 00:00:00",
        last_seen: "2024-01-01 00:00:00",
      },
    ];
    const table = formatTable(freq);
    assert.ok(table.includes("Level"));
    assert.ok(table.includes("Source"));
    assert.ok(table.includes("Message"));
    assert.ok(table.includes("Count"));
    assert.ok(table.includes("First Seen"));
    assert.ok(table.includes("Last Seen"));
  });

  it("includes data values", () => {
    const freq = [
      {
        level: "ERROR",
        source: "auth",
        message: "Login failed",
        count: 5,
        first_seen: "2024-01-01 08:00:00",
        last_seen: "2024-01-01 10:00:00",
      },
    ];
    const table = formatTable(freq);
    assert.ok(table.includes("auth"));
    assert.ok(table.includes("Login failed"));
    assert.ok(table.includes("5"));
  });

  it("has border rows starting with +", () => {
    const freq = [
      {
        level: "ERROR",
        source: "x",
        message: "y",
        count: 1,
        first_seen: "ts",
        last_seen: "ts",
      },
    ];
    const lines = formatTable(freq).split("\n");
    assert.ok(lines[0].startsWith("+"));
    assert.ok(lines[lines.length - 1].startsWith("+"));
  });
});

// ---------------------------------------------------------------------------
// Cycle 8: Full analysis pipeline with JSON output
// ---------------------------------------------------------------------------
describe("analyzeLogFile", () => {
  it("analyses sample.log correctly", () => {
    const result = analyzeLogFile(path.join(FIXTURES, "sample.log"));
    assert.equal(result.total_lines_parsed, 15);
    assert.ok(result.errors_and_warnings > 0);
    assert.ok(result.unique_error_types > 0);
    assert.ok(Array.isArray(result.frequency_table));
  });

  it("top error in sample.log is auth failed login (count=3)", () => {
    const result = analyzeLogFile(path.join(FIXTURES, "sample.log"));
    const top = result.frequency_table[0];
    assert.equal(top.source, "auth");
    assert.equal(top.count, 3);
  });

  it("writes valid JSON output file", () => {
    const tmpPath = path.join(
      os.tmpdir(),
      `log_test_${Date.now()}.json`
    );
    try {
      analyzeLogFile(path.join(FIXTURES, "sample.log"), tmpPath);
      const data = JSON.parse(fs.readFileSync(tmpPath, "utf-8"));
      assert.ok("frequency_table" in data);
      assert.ok("total_lines_parsed" in data);
      assert.ok(Array.isArray(data.frequency_table));
    } finally {
      if (fs.existsSync(tmpPath)) fs.unlinkSync(tmpPath);
    }
  });

  it("handles empty log file", () => {
    const result = analyzeLogFile(path.join(FIXTURES, "empty.log"));
    assert.equal(result.total_lines_parsed, 0);
    assert.equal(result.errors_and_warnings, 0);
    assert.deepStrictEqual(result.frequency_table, []);
  });

  it("handles malformed log gracefully", () => {
    const result = analyzeLogFile(path.join(FIXTURES, "malformed.log"));
    assert.equal(result.total_lines_parsed, 3);
    assert.equal(result.errors_and_warnings, 3);
  });

  it("throws for missing file", () => {
    assert.throws(() => analyzeLogFile("/no/such/file.log"));
  });
});

// ---------------------------------------------------------------------------
// Cycle 9: Edge cases and robustness
// ---------------------------------------------------------------------------
describe("edge cases", () => {
  it("JSON with extra fields still parses", () => {
    const line = JSON.stringify({
      timestamp: "2024-01-15T08:00:00Z",
      level: "ERROR",
      source: "x",
      message: "y",
      extra: 42,
    });
    const result = parseJsonLine(line);
    assert.ok(result !== null);
    assert.equal(result.level, "ERROR");
  });

  it("CRITICAL level parses but is filtered out", () => {
    const line = "2024-01-15 08:00:00 CRITICAL [core] System meltdown";
    const entry = parseSyslogLine(line);
    assert.ok(entry !== null);
    assert.equal(entry.level, "CRITICAL");
    // CRITICAL is not ERROR or WARNING, so should be filtered out
    const filtered = filterErrorsAndWarnings([entry]);
    assert.equal(filtered.length, 0);
  });

  it("single entry frequency table has matching first/last seen", () => {
    const entries = [
      {
        level: "ERROR",
        timestamp: "2024-01-15 08:00:00",
        source: "s",
        message: "m",
      },
    ];
    const freq = buildFrequencyTable(entries);
    assert.equal(freq.length, 1);
    assert.equal(freq[0].first_seen, freq[0].last_seen);
  });

  it("table with multiple rows has correct line count", () => {
    const freq = [
      {
        level: "ERROR",
        source: "a",
        message: "m1",
        count: 5,
        first_seen: "t1",
        last_seen: "t2",
      },
      {
        level: "WARNING",
        source: "b",
        message: "m2",
        count: 2,
        first_seen: "t3",
        last_seen: "t4",
      },
      {
        level: "ERROR",
        source: "c",
        message: "m3",
        count: 1,
        first_seen: "t5",
        last_seen: "t5",
      },
    ];
    const lines = formatTable(freq).split("\n");
    // 3 separator + 1 header + 3 data = 7
    assert.equal(lines.length, 7);
  });
});

File: results/2026-04-07_225702/tasks/02-log-file-analyzer/default-opus/generated-code/test_log_analyzer.py
"""Tests for log file analyzer — written using red/green TDD methodology."""

import json
import os
import pytest
from pathlib import Path

FIXTURES_DIR = Path(__file__).parent / "fixtures"


# ── TDD Round 1: Parse syslog-style lines ──────────────────────────────────

class TestParseSyslog:
    """First piece of functionality: extract level and message from syslog lines."""

    def test_parse_error_line(self):
        from log_analyzer import parse_syslog_line

        line = 'Mar 15 10:22:01 webserver01 nginx[1234]: ERROR: connection refused to upstream backend 10.0.1.5:8080'
        result = parse_syslog_line(line)
        assert result is not None
        assert result["level"] == "ERROR"
        assert result["message"] == "connection refused to upstream backend 10.0.1.5:8080"
        assert result["timestamp"] == "Mar 15 10:22:01"

    def test_parse_warning_line(self):
        from log_analyzer import parse_syslog_line

        line = 'Mar 15 10:22:05 webserver01 nginx[1234]: WARNING: upstream response time exceeded threshold (3.2s)'
        result = parse_syslog_line(line)
        assert result is not None
        assert result["level"] == "WARNING"
        assert result["message"] == "upstream response time exceeded threshold (3.2s)"

    def test_parse_info_line(self):
        from log_analyzer import parse_syslog_line

        line = 'Mar 15 10:26:00 webserver01 nginx[1234]: INFO: health check passed'
        result = parse_syslog_line(line)
        assert result is not None
        assert result["level"] == "INFO"

    def test_returns_none_for_non_syslog(self):
        from log_analyzer import parse_syslog_line

        line = '{"timestamp": "2026-03-15T10:24:00Z", "level": "ERROR"}'
        result = parse_syslog_line(line)
        assert result is None


# ── TDD Round 2: Parse JSON-structured lines ────────────────────────────────

class TestParseJsonLine:
    """Parse JSON log entries and extract level/message/timestamp."""

    def test_parse_json_error(self):
        from log_analyzer import parse_json_line

        line = '{"timestamp": "2026-03-15T10:24:00Z", "level": "ERROR", "service": "auth-api", "message": "connection refused"}'
        result = parse_json_line(line)
        assert result is not None
        assert result["level"] == "ERROR"
        assert result["message"] == "connection refused"
        assert result["timestamp"] == "2026-03-15T10:24:00Z"

    def test_parse_json_warning(self):
        from log_analyzer import parse_json_line

        line = '{"timestamp": "2026-03-15T10:27:00Z", "level": "WARNING", "service": "payment-svc", "message": "slow query"}'
        result = parse_json_line(line)
        assert result["level"] == "WARNING"

    def test_returns_none_for_syslog(self):
        from log_analyzer import parse_json_line

        line = 'Mar 15 10:22:01 webserver01 nginx[1234]: ERROR: something'
        result = parse_json_line(line)
        assert result is None

    def test_returns_none_for_missing_fields(self):
        from log_analyzer import parse_json_line

        line = '{"timestamp": "2026-03-15T10:24:00Z"}'
        result = parse_json_line(line)
        assert result is None


# ── TDD Round 3: Unified parser + file reader with error/warning filter ──────

class TestParseLogFile:
    """Parse a full log file and return only error/warning entries."""

    def test_parse_line_tries_both_formats(self):
        from log_analyzer import parse_line

        syslog = 'Mar 15 10:22:01 webserver01 nginx[1234]: ERROR: fail'
        assert parse_line(syslog)["level"] == "ERROR"

        jsonl = '{"timestamp": "2026-03-15T10:24:00Z", "level": "WARNING", "service": "x", "message": "slow"}'
        assert parse_line(jsonl)["level"] == "WARNING"

        assert parse_line("random garbage line") is None

    def test_parse_log_file_filters_errors_and_warnings(self):
        from log_analyzer import parse_log_file

        # The fixture has 12 lines total: 2 INFO lines should be excluded
        entries = parse_log_file(FIXTURES_DIR / "sample.log")
        levels = {e["level"] for e in entries}
        assert levels == {"ERROR", "WARNING"}
        # Fixture has 7 ERROR + 3 WARNING = 10 error/warning entries
        assert len(entries) == 10

    def test_parse_log_file_nonexistent(self):
        from log_analyzer import parse_log_file

        with pytest.raises(FileNotFoundError):
            parse_log_file(Path("/nonexistent/path.log"))

    def test_parse_log_file_empty(self, tmp_path):
        from log_analyzer import parse_log_file

        empty = tmp_path / "empty.log"
        empty.write_text("")
        assert parse_log_file(empty) == []


# ── TDD Round 4: Frequency table with first/last timestamps ─────────────────

class TestBuildFrequencyTable:
    """Build a frequency table keyed by error message, tracking count + timestamps."""

    def test_basic_frequency_table(self):
        from log_analyzer import build_frequency_table

        entries = [
            {"level": "ERROR", "message": "connection refused", "timestamp": "T1"},
            {"level": "ERROR", "message": "connection refused", "timestamp": "T3"},
            {"level": "WARNING", "message": "slow query", "timestamp": "T2"},
        ]
        table = build_frequency_table(entries)
        assert len(table) == 2
        assert table["connection refused"]["count"] == 2
        assert table["connection refused"]["level"] == "ERROR"
        assert table["connection refused"]["first_seen"] == "T1"
        assert table["connection refused"]["last_seen"] == "T3"
        assert table["slow query"]["count"] == 1
        assert table["slow query"]["first_seen"] == "T2"
        assert table["slow query"]["last_seen"] == "T2"

    def test_empty_entries(self):
        from log_analyzer import build_frequency_table

        assert build_frequency_table([]) == {}

    def test_fixture_file_frequency(self):
        """Integration test: frequency table from the sample fixture."""
        from log_analyzer import parse_log_file, build_frequency_table

        entries = parse_log_file(FIXTURES_DIR / "sample.log")
        table = build_frequency_table(entries)
        # "connection refused to upstream backend 10.0.1.5:8080" appears 3 times
        conn_err = table["connection refused to upstream backend 10.0.1.5:8080"]
        assert conn_err["count"] == 3
        assert conn_err["level"] == "ERROR"
        # "timeout waiting for database connection" appears 2 times
        assert table["timeout waiting for database connection"]["count"] == 2


# ── TDD Round 5: Output formatting — table string and JSON export ────────────

class TestOutputFormatting:
    """Format the frequency table as a human-readable string and write JSON."""

    def test_format_table_string(self):
        from log_analyzer import format_table

        table = {
            "connection refused": {
                "level": "ERROR", "count": 3,
                "first_seen": "T1", "last_seen": "T5",
            },
            "slow query": {
                "level": "WARNING", "count": 1,
                "first_seen": "T2", "last_seen": "T2",
            },
        }
        output = format_table(table)
        # Should contain header and both rows
        assert "connection refused" in output
        assert "slow query" in output
        assert "ERROR" in output
        assert "WARNING" in output
        # Count should appear
        assert "3" in output
        assert "1" in output

    def test_format_table_sorted_by_count_desc(self):
        from log_analyzer import format_table

        table = {
            "rare error": {"level": "ERROR", "count": 1, "first_seen": "T1", "last_seen": "T1"},
            "common error": {"level": "ERROR", "count": 10, "first_seen": "T1", "last_seen": "T9"},
        }
        output = format_table(table)
        # "common error" (count=10) should appear before "rare error" (count=1)
        assert output.index("common error") < output.index("rare error")

    def test_write_json_report(self, tmp_path):
        from log_analyzer import write_json_report

        table = {
            "err1": {"level": "ERROR", "count": 2, "first_seen": "T1", "last_seen": "T2"},
        }
        out_path = tmp_path / "report.json"
        write_json_report(table, out_path)
        data = json.loads(out_path.read_text())
        assert len(data["errors"]) == 1
        assert data["errors"][0]["message"] == "err1"
        assert data["errors"][0]["count"] == 2
        assert "total_entries" in data["summary"]


# ── TDD Round 6: CLI end-to-end integration ──────────────────────────────────

class TestCLI:
    """Test the main() CLI entry point end-to-end."""

    def test_analyze_fixture_produces_both_outputs(self, tmp_path, capsys):
        from log_analyzer import analyze

        json_out = tmp_path / "report.json"
        analyze(FIXTURES_DIR / "sample.log", json_out)

        # Human-readable table was printed to stdout
        captured = capsys.readouterr().out
        assert "connection refused" in captured
        assert "ERROR" in captured

        # JSON report was written
        assert json_out.exists()
        data = json.loads(json_out.read_text())
        assert data["summary"]["total_entries"] == 10
        assert data["summary"]["unique_messages"] == 5

    def test_cli_missing_file(self, capsys):
        """Graceful error message for missing files instead of a traceback."""
        from log_analyzer import analyze

        with pytest.raises(SystemExit):
            analyze(Path("/does/not/exist.log"), Path("/tmp/out.json"))

    def test_cli_empty_file(self, tmp_path, capsys):
        from log_analyzer import analyze

        empty = tmp_path / "empty.log"
        empty.write_text("")
        json_out = tmp_path / "report.json"
        analyze(empty, json_out)

        captured = capsys.readouterr().out
        assert "No errors or warnings found" in captured
